nat-lab/tests/test_batching.py

```diff
@@ -0,0 +1,248 @@
+        node_container_names = [
+            conn.container_name()
+            for conn in conns
+            if isinstance(conn, DockerConnection)
+        ]
+
+        container_names = gateway_container_names + node_container_names
+        # TODO: assert these are docker connections
+        print("Will capture batching on containers: ", container_names)
+        cnodes = zip(env.clients, env.nodes, [c.connection for c in env.connections])
+
+        # the plan is to misalign the peers for better batcing observation. WIthout misalignment
+        # peers will form a meshnet almost immediately, already achieving the desired batching effect.
+        for client in env.clients:
+            await client.stop_device()
+
+        tasks = []
+        for name in container_names:
+            task = asyncio.create_task(
+                capture_traffic(
+                    name,
+                    capture_duration,
+                )
+            )
+            tasks.append(task)
+
+        api = env.api
+
+        # misalign the peers by sleeping some before starting each node again
+        for client, node, _ in cnodes:
+            await asyncio.sleep(6)
+            await client.simple_start()
+            await client.set_meshmap(api.get_meshmap(node.id))
+
+        await asyncio.gather(*(
+            client.wait_for_state_peer(
+                peer_node.public_key, [State.Connected], [PathType.Direct]
+            )
+            for client, peer_node in product(clients, peer_nodes)
+            if not client.is_node(peer_node)
+        ))
+
+        pcap_paths = await asyncio.gather(*tasks)
+
+        # pylint: disable=consider-using-set-comprehension
+        pyro5_ports = list(set([client.get_proxy_port() for client in env.clients]))
+        assert len(pyro5_ports) == 1
+        pyro5_port = pyro5_ports[0]
+
+        filters = [
+            None,
+            ("icmpv6echorequest", lambda p: p.haslayer(ICMPv6EchoRequest)),
+            (
+                "ipv46 non-pyro5",
+                lambda p: (
+                    (p.haslayer(IP) or p.haslayer(IPv6))
+                    and pyro5_port not in (p[TCP].sport, p[TCP].dport)
+                    if p.haslayer(TCP)
+                    else True
+                ),
+            ),
+        ]
+
+        for container, pcap_path in zip(container_names, pcap_paths):
+            for filt in filters:
+                filter_name = filt[0] if filt else "none"
+                hs = generate_histogram_from_pcap(
+                    pcap_path, capture_duration, 1, filt[1] if filt else None
+                )
+                title = (
+                    f"{container}-filter({filter_name})-id({request.node.callspec.id})"
+                )
+                draw_histogram(title, hs, max_height=12)
+
+        for p in pcap_paths:
+            os.unlink(p)
+
+        assert (
+            False
+        ), "TODO: implement a proper assert(this assert also forced the dump of output)"
```

Comment:
So, as I understand this test will always fail on this line - I guess it would break CI when merged, soo, what is the plan to do with it? Mark it as flaky or sth like that?

Reply:
Good point. So I wanted the quickest way to see it in action, hence I just `print` stuff. Printing works fine and it's visible when the test fails, so I need the failure while we work on it.

If I put it in the nightly, it will always fail and I am not sure if that's a good thing, might be an annoyance, instead I plan immediately afterwards to introduce another job into the `libtelio-build` pipeline which would also run nightly just for this.

With no assert, `pytest` assumes it all went find and doesn't emit the captured stdout and stderr.

An alternative is possible to save those diagrams to files or something like that however this PR already feels like walking through a mud so I opted out for simplicity

Reply:
I reworked this one and now the assert is gone. It rather renders histograms into textfiles at `logs/`.

Reply:
I think the `report.html` still has the output even for passed tests. Printing to files is also more than fine